File: src/main/java/me/lovesasuna/lanzou/file/FolderImpl.java
package me.lovesasuna.lanzou.file;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import me.lovesasuna.lanzou.bean.Triple;
import me.lovesasuna.lanzou.util.BasicUtil;
import me.lovesasuna.lanzou.util.NetWorkUtil;
import me.lovesasuna.lanzou.util.ReaderUtil;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * @author LovesAsuna
 * @date 2020/8/22 10:31
 **/
public class FolderImpl extends FolderItem {

    public FolderImpl(String id) {
        this.id = id;
    }

    @Override
    public void init(BufferedReader reader) {
        try {
            // 文件夹
            Triple<Integer, InputStream, Long> result;
            ReaderUtil.readAnyTime(68, reader);
            String t = reader.readLine().split("'")[1];
            String k = reader.readLine().split("'")[1];
            ReaderUtil.readAnyTime(9, reader);
            String lx = "2";
            String fid = "" + BasicUtil.extractInt(reader.readLine(), 0);
            String uid = "" + BasicUtil.extractInt(reader.readLine(), 0);
            String pg = "1";
            String rep = "0";
            String up = "1";
            reader.close();
            StringBuilder builder = new StringBuilder();
            builder.append("lx=").append(lx).append("&")
                    .append("fid=").append(fid).append("&")
                    .append("uid=").append(uid).append("&")
                    .append("pg=").append(pg).append("&")
                    .append("rep=").append(rep).append("&")
                    .append("t=").append(t).append("&")
                    .append("k=").append(k).append("&")
                    .append("up=").append(up);
            result = NetWorkUtil.post("https://wwa.lanzous.com/filemoreajax.php", builder.toString().getBytes());
            Objects.requireNonNull(result);
            reader = new BufferedReader(new InputStreamReader(result.second));
            ObjectMapper mapper = new ObjectMapper();
            JsonNode root = mapper.readTree(reader.readLine());
            reader.close();
            JsonNode text = root.get("text");
            int length = text.size();
            fileItemSet = new HashSet<>();
            for (int i = 0; i < length; i++) {
                JsonNode fileNode = text.get(i);
                String id = fileNode.get("id").asText();
                FileItem file = new FileImpl(id);
                file.icon = fileNode.get("icon").asText();
                file.name = fileNode.get("name_all").asText();
                file.size = fileNode.get("size").asText();
                file.time = fileNode.get("time").asText();

                String lanzousUrl = "https://wwa.lanzous.com/" + id;
                result = NetWorkUtil.get(lanzousUrl);
                Objects.requireNonNull(result);
                reader = new BufferedReader(new InputStreamReader(result.second));
                if (Item.isFile(reader)) {
                    file.init(reader);
                }
                fileItemSet.add(file);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean download(Path path) {
        AtomicBoolean success = new AtomicBoolean(true);
        fileItemSet.forEach(fileItem -> {
            success.set(success.get() && fileItem.download(path));
        });
        return success.get();
    }
}

File: src/main/kotlin/me/lovesasuna/lanzou/file/FolderImpl.kt
package me.lovesasuna.lanzou.file

import com.fasterxml.jackson.databind.ObjectMapper
import me.lovesasuna.lanzou.util.BasicUtil
import me.lovesasuna.lanzou.util.NetWorkUtil
import me.lovesasuna.lanzou.util.ReaderUtil
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.file.Path
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Consumer

/**
 * @author LovesAsuna
 * @date 2020/8/22 10:31
 */
class FolderImpl(id: String) : FolderItem() {
    override fun init(reader: BufferedReader) {
        var reader = reader
        try {
            // 文件夹
            var result: Triple<Int, InputStream, Long>?
            ReaderUtil.readAnyTime(68, reader)
            val t = reader.readLine().split("'").toTypedArray()[1]
            val k = reader.readLine().split("'").toTypedArray()[1]
            ReaderUtil.readAnyTime(9, reader)
            val lx = "2"
            val fid = "" + BasicUtil.extractInt(reader.readLine(), 0)
            val uid = "" + BasicUtil.extractInt(reader.readLine(), 0)
            val pg = "1"
            val rep = "0"
            val up = "1"
            reader.close()
            val builder = StringBuilder()
            builder.append("lx=").append(lx).append("&")
                .append("fid=").append(fid).append("&")
                .append("uid=").append(uid).append("&")
                .append("pg=").append(pg).append("&")
                .append("rep=").append(rep).append("&")
                .append("t=").append(t).append("&")
                .append("k=").append(k).append("&")
                .append("up=").append(up)
            result = NetWorkUtil.post("https://wwa.lanzous.com/filemoreajax.php", builder.toString().toByteArray())
            Objects.requireNonNull(result)
            reader = BufferedReader(InputStreamReader(result!!.second))
            val mapper = ObjectMapper()
            val root = mapper.readTree(reader.readLine())
            reader.close()
            val text = root["text"]
            val length = text.size()
            fileItemSet = HashSet()
            for (i in 0 until length) {
                val fileNode = text[i]
                val id = fileNode["id"].asText()
                val file: FileItem = FileImpl(id)
                file.icon = fileNode["icon"].asText()
                file.name = fileNode["name_all"].asText()
                file.size = fileNode["size"].asText()
                file.time = fileNode["time"].asText()
                val lanzousUrl = "https://wwa.lanzous.com/$id"
                result = NetWorkUtil[lanzousUrl]
                Objects.requireNonNull(result)
                reader = BufferedReader(InputStreamReader(result!!.second))
                if (isFile(reader)) {
                    file.init(reader)
                }
                fileItemSet.add(file)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun download(path: Path): Boolean {
        val success = AtomicBoolean(true)
        fileItemSet.forEach(Consumer { fileItem: FileItem -> success.set(success.get() && fileItem.download(path)) })
        return success.get()
    }

    init {
        this.id = id
    }
}